app/controllers/admin_controller.rb

```diff
@@ -123,6 +123,38 @@ def mark_spam
     end
   end
 
+  def mark_comment_spam
+    @comment = Comment.find params[:id]
+    if current_user && (current_user.role == 'moderator' || current_user.role == 'admin')
+      if @comment.status != 2
+        @comment.spam
+        flash[:notice] = "Comment has been marked as spam."
+      else
+        flash[:notice] = "Comment already marked as spam."
+      end
+    else
+      flash[:error] = 'Only moderators can moderate comments.'
+    end
+    redirect_to '/dashboard'
+  end
+
+  def publish_comment
+    if current_user && (current_user.role == 'moderator' || current_user.role == 'admin')
+      @comment = Comment.find params[:id]
+      if @comment.status != 2
+        flash[:notice] = 'Comment already published.'
+      else
+        @comment.publish
```

Comment:
Or maybe you can instead check here if `status != 0`, and if `true` set it to 0. This would be parallel to the strategy you used above to mark a comment as spam.

Reply:
These are comments on clarity, not correctness. I think the code itself has no problem.

Reply:
thanks, @seafr I really appreciate your comments and will correct this.